### ISSUE TITLE
TALON

### ISSUE THREAD

**Issue by deGENERATIVE-SQUAD** [Clybius/Personalized-Optimizers#1]
Hello.

TALON is very good, excellent work.

Could you answer a few questions:

1. Adaptation is somewhat slow and requires high learning rates (around 1e-3 or even 1e-2) for noticeable speed, especially when using Huber + SNR (L2 doesn't require such high values). How can adaptation be accelerated without increasing the learning rate?

2. Am I correct in understanding that, given the claimed adaptiveness, one can simply use a constant learning rate scheduler?

3. Does it make sense to use EDM2 together with TALON?

4. The adaptive ability is quite interesting, I haven’t seen anything like it before: I observe a literal morphing of generation toward the dataset progressively, without critical deviations. For example, with Prodigy, I see the optimizer trying to "guess" the next favorable direction by trial and error, grabbing concepts from the dataset while partially ignoring the base model, almost copying the dataset directly. TALON, on the other hand, seems to always use the base model as a guide. This makes result prediction and choosing the right epoch much simpler compared to other optimizers - you know that as long as the generations don’t start producing something incoherent, the training is on track. What mechanism enables this kind of behavior?

**Comment by Clybius** [comment]
Hiya! Thank you for the kind words.

1. This is likely due to a given parameter's sign flipping from positive to negative, resulting in a low multiplier on the end gradient. I can implement a power scalar value for the sign's momentum soon, which would default to 1.0. Thus if you want it to be faster, you lower this power scalar to 0.5 or even lower (with 0 disabling sign-momentum scaling entirely). This should make speed a bit faster when there's a medium amount of confidence in the direction/sign of a gradient's parameter.

2. The adaptive-ness in TALON is moreso akin to Adam's adaptive-ness, in that the gradient is pre-conditioned to be scale-invariant (normalized to around 1.0) via a squared momentum denominator. I'd still recommend using a learning rate that decays at the end of training like a trapezoidal LR or warmup_stable_decay, but it's not necessary. Optimally, in noise-less environments due to how the sign momentum is implemented, the gradient's value is multiplied to near 0. This is because we remove the sign from the gradient for further calculations, and then momentumize the sign via a lerp to -1 (if negative) or 1 (if positive). 
If you're near the minima in such a training environment, the sign will jump from negative to positive, which naturally decays to a near-0 value due to the jumping between -1 and 1 at a weight of 0.1 (as the default beta is 0.9).

3. Sure! Practically any optimizer should work fine with EDM2, some just work better than others. TALON seems to be good with it from my SDXL testing.

4. What you're seeing with Prodigy is likely due to how Prodigy was designed. It's not 'smooth' like Adam-style optimizers, where the gradient is put into a momentum state to smooth out higher frequencies/jumps in the gradient from step-to-step. Prodigy can be more thought of like RMSprop in the sense of how its updates work: The gradient is passed through a denominator (RMSprop) and does calculations on the `p` state (parameters) to then determine a given learning rate, if I recall correctly.

I'd attribute the better training samples to the momentum states' ability to smooth out high frequency jumps in magnitude and the separation of sign from the magnitude. The gradient is even 'smoother' than Adam due to TALON's second beta, or 'value momentum', being longer/slower. 
For example, in Adam, the default momentum beta is set to 0.9. In TALON, the equivalent beta is set to 0.99. There's also the fact that we're calculating momentum via `grad.abs()`, or the magnitude of the gradient. In Adam, the gradient is momentumized by just `grad`. Here's an example of how this helps:
In Adam, a lerp from the value of 1 to the value of -1 at a weight of 0.1 (due to the 0.9 beta,) will result in a value of 0.8.
In TALON, a lerp from the value of 1 to the value of |-1| at a weight of 0.01 (due to the 0.99 beta,) will result in a value of 1.0, i.e. no change.

Hope this helps, even though its a little technical!

**Comment by Clybius** [comment]
Leaving another comment to let you know that I've introduced a new parameter called `signscale_power`. A value of 0.5 may help increase speed (default 1.0), as less 'confidence' is needed in the sign in order to scale it up to its un-scaled self.

If that doesn't help, try to set both `spectral_min` and `spectral_max` to 0, and set `spectral_adaptive` to `True`. This enables general Newton-Schulz iterations with scale adaptivity.

**Comment by deGENERATIVE-SQUAD** [comment]
@Clybius 
Got it, made some tests.

>A value of 0.5 may help increase speed (default 1.0), as less 'confidence' is needed in the sign in order to scale it up to its un-scaled self.

0.5 differs very little from 1 in my test. I set it to 0.1 with a learning rate of 1e-3 - training became generally effective, but unstable. Switched to 1e-4 and the changes became negligible again. Changed to 5e-4 - it's stable and faster than usual, but not as fast as the extreme 1e-3.

>If that doesn't help, try to set both spectral_min and spectral_max to 0, and set spectral_adaptive to True. This enables general Newton-Schulz iterations with scale adaptivity.

Thanks, will check.

**Comment by deGENERATIVE-SQUAD** [comment]
UPD. The issue turned out to be with the L2/L1/Huber losses. Everything works great with the wavelet loss.

**Comment by Clybius** [comment]
Hmm, it should still work fine on those losses. 

Just had a thought, you could also perhaps change the last beta to something lower like 0.99 or 0.999. This will make it so that the max denominator length is 0.99 or 0.999 instead of growing to 1 - 1e-7. 

The last beta is intended to be a denominator that is an average of every sample, but it may help in some cases to lower it to an Adam-like value.

**Comment by deGENERATIVE-SQUAD** [comment]
@Clybius 
>Hmm, it should still work fine on those losses

That’s quite possible with regular LoRA, but in this case, I’m training BOFT.

>Just had a thought, you could also perhaps change the last beta to something lower like 0.99 or 0.999. 

Will try, thanks
